src/CloudNative.CloudEvents/CloudEventAttribute.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright 2021 Cloud Native Foundation. 
+// Licensed under the Apache 2.0 license.
+// See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace CloudNative.CloudEvents
+{
+    // TODO: Allow additional validation, e.g. for extensions and content type?
+    // TODO: Add Format/Parse methods that pass through to the type (and do validation when parsing)?
+
+    /// <summary>
+    /// An attribute that can be associated with a <see cref="CloudEvent"/>.
+    /// This may be a context attribute or an extension attribute.
+    /// This class represents the abstract concept of an attribute, rather than
+    /// an attribute value.
+    /// </summary>
+    public class CloudEventAttribute
+    {
+        /// <summary>
+        /// The type of the attribute. All values provided must be compatible with this.
+        /// </summary>
+        public CloudEventAttributeType Type { get; }
+
+        /// <summary>
+        /// The name of the attribute. Instances of this class associated with different
+        /// versions of the specification may use different names for the same concept.
+        /// </summary>
+        public string Name { get; }
+
+        /// <summary>
+        /// Indicates whether this attribute is a required attribute.
+        /// Extension attributes are never required.
+        /// </summary>
+        public bool IsRequired { get; }
+
+        /// <summary>
+        /// Indicates whether this attribute is an extension attribute.
+        /// </summary>
+        public bool IsExtension { get; }
+
+        private Action<object> validator;
+
+        // TODO: Have a "mode" of Required/Optional/Extension?
```

Comment:
Wouldn't an extension always be optional though?

Reply:
Yes - it would just be potentially simpler to have a single "mode" instead of two separate (but not entirely orthogonal) boolean values.

I'm definitely happy to wait and see whether it turns out it would actually be useful.

Reply:
Yeah I don't have a strong preference here. Maybe we could make it more clear that required and optional are both for reserved attributes. RequiredReserved, OptionalReserved, Extension?

Reply:
Hmm... maybe. Let's work out naming if and when we decide that the enum is worth doing :)